libsolidity/codegen/ExpressionCompiler.cpp

```diff
@@ -1282,35 +1304,54 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
 
 			if (hasSelectorOrSignature)
 			{
-				// stack: <selector> <memory pointer>
-				solAssert(arguments.size() >= 1, "");
-				Type const* selectorType = arguments[0]->annotation().type;
-				utils().moveIntoStack(selectorType->sizeOnStack());
-				Type const* dataOnStack = selectorType;
-				// stack: <memory pointer> <selector>
-				if (function.kind() == FunctionType::Kind::ABIEncodeWithSignature)
+				Type const* dataOnStack = nullptr;
+
+				if (
+					function.kind() == FunctionType::Kind::ABIEncodeCall &&
+					functionPtr->hasDeclaration()
+				)
 				{
-					// hash the signature
-					if (auto const* stringType = dynamic_cast<StringLiteralType const*>(selectorType))
+					m_context << util::selectorFromSignature(functionPtr->externalSignature());
+					dataOnStack = TypeProvider::fixedBytes(4);
+				}
+				else
+				{
+					// stack: <selector> <memory pointer>
```

Comment:
This comment is wrong ATM

Reply:
Then the comment in line 1303 is also wrong. Why do you think it's wrong?

Reply:
Because in case of `abi.encodeCall`, the stack contains a full function pointer and not just a selector.

Reply:
hmm I see. Then it was previously also wrong for the case of `EncodeWithSignature`, it just got one more meaning now. 
We can correct that by calling it .. well, I struggle to come up with a generic enough name. `selector-ish`?

Reply:
`selector/functionPointer/signature`